File: python/euler/bin/p0004.py
# project euler: problem 4

from euler.lib.util import is_palindrome


def compute(digits: int) -> str:
    assert digits > 0, 'range error'
    n_upper = (10**digits) - 1
    n_lower = 10 ** (digits - 1)
    blk_upper_limit = 10 ** (digits * 2) - 1
    blk_lower_limit = 10 ** ((digits - 1) * 2) if digits > 1 else 0
    blk_width = 10 ** (digits * 2 - 2)
    answer = []

    # The blocks of product of two numbers are checked in descending order
    # The block width, blk_width, is 1/100 of the upper limit
    for blk_lower in reversed(range(blk_lower_limit, blk_upper_limit, blk_width)):
        blk_upper = blk_lower + blk_width - 1
        for x in reversed(range(n_lower, n_upper + 1)):
            if x * x < blk_lower:
                break
            for y in reversed(range(n_lower, min(blk_upper // x, x) + 1)):
                if (tmp := x * y) < blk_lower:
                    break
                elif is_palindrome(tmp):
                    answer.append(tmp)

        if answer:
            return str(max(answer))

    assert False, 'unreachable!'


def solve() -> str:
    return compute(3)  # 3-digit

File: typescript/euler/bin/p0004.ts
// project euler: problem 4

import { maxLst } from "../lib/math.ts";
import { isPalindrome, range } from "../lib/util.ts";

export const compute = (digits: number): string => {
  if (digits <= 0) {
    throw new Error("range error");
  }

  const nUpper = (10 ** digits) - 1;
  const nLower = 10 ** (digits - 1);
  const blkUpperLimit = 10 ** (digits * 2) - 1;
  const blkLowerLimit = digits > 1 ? 10 ** ((digits - 1) * 2) : 0;
  const blkWidth = 10 ** (digits * 2 - 2);
  const answer: number[] = [];

  for (
    const blk_lower of range(blkLowerLimit, blkUpperLimit, blkWidth)
      .toReversed()
  ) {
    const blk_upper = blk_lower + blkWidth - 1;
    for (const x of range(nLower, nUpper + 1).toReversed()) {
      if (x * x < blk_lower) {
        break;
      }
      for (
        const y of range(nLower, Math.min(Math.trunc(blk_upper / x), x) + 1)
          .toReversed()
      ) {
        const tmp = x * y;
        if (tmp < blk_lower) {
          break;
        }
        if (isPalindrome(tmp)) {
          answer.push(tmp);
        }
      }
    }
    if (answer.length !== 0) {
      break;
    }
  }

  return answer.length !== 0 ? String(maxLst(answer)) : "0";
};

export const solve = (): string => compute(3); // 3-digit
